Use ComplexTextController even for the simple text code path for GTK and WPE ports

fujii opened the pull request:
#### 23619bf93e9cbd28c08b0410deb4c72ebe1337a5
<pre>
Use ComplexTextController even for the simple text code path for GTK and WPE ports
<a href="https://bugs.webkit.org/show_bug.cgi?id=297843">https://bugs.webkit.org/show_bug.cgi?id=297843</a>

Reviewed by Nikolas Zimmermann and Vitor Roriz.

&lt;<a href="https://commits.webkit.org/200076@main">https://commits.webkit.org/200076@main</a>&gt; changed GTK and WPE ports to
always use the complex text code path. This was done by using
`FontCascade::setCodePath(CodePath::Complex)`. However, this approach
involved a unnecessary performance regression.

In most of layout perf test cases, the complex and simple text code
paths consume only a tiny portion of time. In WebCore layout and
rendering code, there are some conditions checking the code path like
`if (fontCascade.codePath(TextRun(content)) == FontCascade::CodePath::Complex)`.
This yields the difference in performance. For example, the following
condition in LineBoxBuilder::collectFallbackFonts is a condition
making the difference.

    if (inlineTextBox.canUseSimplifiedContentMeasuring()) {
        // Simplified text measuring works with primary font only.
        return { };
    }

Thus, we shouldn&apos;t use `FontCascade::setCodePath()` to choose the
complex text code path.

To use the simple text code path, we have to implement
Font::applyTransforms() for shaping the text. In bug#297088, I studied
the possibility of implementing Font::applyTransforms with HarfBuzz.
However, HarfBuzz doesn&apos;t have an API to implement
Font::applyTransforms efficiently.

Use ComplexTextController even for the simple text code path for GTK
and WPE ports. Added a new macro ENABLE_COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH.

* LayoutTests/platform/gtk/fast/text/international/pop-up-button-text-alignment-and-direction-expected.txt:
* Source/WTF/wtf/PlatformEnableGlib.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::drawText const):
(WebCore::FontCascade::drawEmphasisMarks const):
(WebCore::FontCascade::displayListForTextRun const):
(WebCore::FontCascade::widthOfTextRange const):
(WebCore::FontCascade::width const):
(WebCore::FontCascade::adjustSelectionRectForText const):
(WebCore::FontCascade::offsetForPosition const):
(WebCore::FontCascade::codePath const):
(WebCore::FontCascade::widthForCharacterInRun const):
(WebCore::FontCascade::lineSegmentsForIntersectionsWithRect const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::widthForTextUsingSimplifiedMeasuring const):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitialize):

Canonical link: <a href="https://commits.webkit.org/299354@main">https://commits.webkit.org/299354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df260e7d752d7fc7ecaeceae770596ba33364036

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124904 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39098 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90104 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121674 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31150 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70610 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30209 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24553 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68563 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110839 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100592 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24741 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127960 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117235 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45628 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34438 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98749 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45992 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102660 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98531 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43973 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21975 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42136 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18917 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51176 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145931 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44962 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37525 "Found 1 new JSC binary failure: testapi, Found 18582 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48308 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46648 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -247,6 +247,18 @@ class FontCascade final : public CanMakeWeakPtr<FontCascade>, public CanMakeChec
 private:
     enum class ForTextEmphasisOrNot : bool { NotForTextEmphasis, ForTextEmphasis };
 
+    CodePath codePathPrivate(const TextRun& run, std::optional<unsigned> from = std::nullopt, std::optional<unsigned> to = std::nullopt) const
```
I would use a more descriptive name, e.g. determineCodePath, and then rename codePath to determineCodePathForTextRun.

nikolaszimmermann replied:
You can also leave out the 'run' parameter name, doesn't add value.

carlosgcampos replied:
Maybe codePathForShaping?

fujii replied:
The variable 'run' is used below. Can't leave it out.

fujii replied:
`FontCascade` has `requiresShaping()` method. `codePathForShaping` doesn't seem to be nice. I think `internal` or `private` is a good word for it.

nikolaszimmermann replied:
Oops sure you need run 🤣

nikolaszimmermann replied:
codePathPrivate is as confusing as codePathInternal IMHO.

fujii replied:
The idea of this change is stating different code paths for the external and internal.
FontCascade::codePath() is for the external, and returning a code path for the external code.
FontCascade::codePathPrivate() is for the internal, and returning a code path for the internal code.
The method name determineCodePathForTextRun() doesn't sound good to me.

fujii replied:
However, FontCascade has two `codePath` methods, `codePath()` and `codePath(const TextRun&, std::optional<unsigned>, std::optional<unsigned>) const`.
I think it's a good idea to rename the latter to `codePathForTextRun(const TextRun&, ...) const`.
But, this is another topic. I don't want to do it in the patch.
IMHO, the `codePath(const TextRun&, ...)` method should be TextRun's method. And, TextRun should cache the result. But, I don't understand enough. I don't suggest to do that.

nikolaszimmermann replied:
You might not be happy with the naming proposal, that is ok:

```
FontCascade::codePath() is for the external, and returning a code path for the external code.
FontCascade::codePathPrivate() is for the internal, and returning a code path for the internal code.
```

But sorry, this is not clear from the naming of the functions. codePath() and codePathPrivate() is non-obvious, and it's not clear to anyone when which code path should be used. Maybe @carlosgcampos has a better suggestion?

carlosgcampos replied:
I still don't see the problem of codePathForShaping(), but I agree Private or Internal are confusing.

fujii replied:
Fixed.

nikolaszimmermann reviewed (commented):
Looks great!

nikolaszimmermann commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -345,6 +345,7 @@ float FontCascade::width(const TextRun& run, SingleThreadWeakHashSet<const Font>
 }
 NEVER_INLINE float FontCascade::widthForSimpleTextSlow(StringView text, TextDirection textDirection, float* cacheEntry) const
 {
+#if !ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH)
```
I think this condition shouldn't be in the `widthForSimpleTextSlow` path, but in the one place that calls this:

```
inline float FontCascade::widthForTextUsingSimplifiedMeasuring(StringView text, TextDirection textDirection) const
{   
    if (text.isEmpty())
        return 0;
    ASSERT(codePath(TextRun(text)) != CodePath::Complex);
    float* cacheEntry = fonts()->widthCache().add(text, std::numeric_limits<float>::quiet_NaN());
    if (cacheEntry && !std::isnan(*cacheEntry))
        return *cacheEntry;
    
    return widthForSimpleTextSlow(text, textDirection, cacheEntry);
}
```

There we should check if the unified code path is used, and then call `widthForComplexText` instead.

fujii replied:
The returned width has to be put into the width cache.

nikolaszimmermann replied:
Yeah sure. you can do that in widthForTextUsingSimplifiedMeasuring?

nikolaszimmermann replied:
I was thinking of:

```
inline float FontCascade::widthForTextUsingSimplifiedMeasuring(StringView text, TextDirection textDirection) const
{   
    if (text.isEmpty())
        return 0;
    ASSERT(codePath(TextRun(text)) != CodePath::Complex);
    float* cacheEntry = fonts()->widthCache().add(text, std::numeric_limits<float>::quiet_NaN());
    if (cacheEntry && !std::isnan(*cacheEntry))
        return *cacheEntry;

#if ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH)
    TextRun run { text, 0, 0, ExpansionBehavior::defaultBehavior(), textDirection, false, false };
    auto newWidth = widthForComplexText(run, nullptr, nullptr);
    if (cacheEntry)
        *cacheEntry = newWidth;
    return newWidth;
#else
    return widthForSimpleTextSlow(text, textDirection, cacheEntry);
#endif
}
```

(untested)

fujii replied:
FontCascade::widthForTextUsingSimplifiedMeasuring is a inline function which is performace critical. https://bugs.webkit.org/show_bug.cgi?id=270612
I don't think your code is better because you add more not-hot path code to the inline function.
You don't like my code because FontCascade::widthForSimpleTextSlow calls widthForComplexText. But, the idea of ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH) is using ComplexTextController in the simple text code path. I think no poblem for my code.

nikolaszimmermann replied:
My proposal doesn't change anything for non-glib ports, so it's perf neutral by definition.
For glib ports it doesn't add any overhead when looking up from the width cache.
When measuring a new TextRun, I find it hard to imagine this adds any overhead, as before you called a non-inlined function (widthForSimpleTextSlow), that called _another_ non-inlined function (widthForComplexText).

With the proposal we only setup a TextRun and call _one_ non-inlined function, and avoid indirection of going first via widthForSimpleTextSlow. So it's likely even faster, than what you had before.

fujii replied:
The original inline function FontCascade::widthForTextUsingSimplifiedMeasuring checks the width cache first. Then, it returns early if the width cache hits. It calls widthForSimpleTextSlow() function for the slow code path (no cache hit). Your suggestion adds more no-hot code to the very hot inline function. It make the inline function bigger.

nikolaszimmermann replied:
> The original inline function FontCascade::widthForTextUsingSimplifiedMeasuring checks the width cache first. Then, it returns early if the width cache hits.

That's exactly the same with my suggestion. Please double-check.

fujii replied:
Your code add extra non-hot code to the hot inline function. This part
> if (cacheEntry)
>        *cacheEntry = newWidth;

nikolaszimmermann replied:
Yes and? It reduces two non-inlined function calls to one - that's for sure cheaper :-)

nikolaszimmermann replied:
And it makes the code easier to grasp, calling widthForComplexText from widthForSimpleTextSlow is unintuitive, and mixes concerns - - the widthForSimpleTextSlow method shouldn't need to know about COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH.

fujii replied:
Fixed.

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
```diff
@@ -143,8 +143,6 @@ static void seatDevicesChangedCallback(GdkSeat* seat, GdkDevice*, WebProcessPool
 {
-    m_alwaysUsesComplexTextCodePath = true;
-
     if (const char* forceComplexText = getenv("WEBKIT_FORCE_COMPLEX_TEXT"))
```
Do we still need this logic and `m_alwaysUsesComplexTextCodePath` ?

fujii replied:
It's still used by the public WK API WKContextSetAlwaysUsesComplexTextCodePath which is used by WebKitTestRunner.

nikolaszimmermann replied:
Ah ok, so let's keep this.

vitorroriz commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -233,6 +233,19 @@ class FontCascade final : public CanMakeWeakPtr<FontCascade>, public CanMakeChec
+
+    CodePath codePathForShaping(const TextRun& run, std::optional<unsigned> from = std::nullopt, std::optional<unsigned> to = std::nullopt) const
+    {
+#if ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH)
+        UNUSED_PARAM(run);
+        UNUSED_PARAM(from);
+        UNUSED_PARAM(to);
+        return CodePath::Complex;
+#else
```
Any reason why we don't simply add this fragment to codePath itself?

fujii replied:
That's the current approach. GTK and WPE ports are always choosing the complex text code path by doing `FontCascade::setCodePath(CodePath::Complex)`.

vitorroriz replied:
I'm saying that this PR creates a new function `codePathForShaping` that will either return `CodePath::Complex;` or call the original `codePath`. So why can't codePath just added this fragment directly?

fujii replied:
I don't understand your idea. I need both methods. `codePath(const TextRun&, ...)`  is used for WebCore layout and rendering code. `codePathForShaping(const TextRun&, ..)` is used for FontCascade class.

nikolaszimmermann replied:
@vitorroriz - it's a bit confusing I agree. Let me try to elaborate a bit:

The point is that @fujii wants `canUseSimplifiedTextMeasuring` to return true for text that would identify as "simple text":

```
static bool canUseSimplifiedTextMeasuring(StringView content, const FontCascade& fontCascade, bool whitespaceIsCollapsed)
{
    if (fontCascade.codePath(TextRun(content)) == FontCascade::CodePath::Complex)
        return false;
...
```

If `codePath()` itself would return `CodePath::Complex`, then we would always `return false` in `canUseSimplifiedTextMeasuring` -- making our ports _always_ use the non-simplified text measuring code paths in LayoutTreeBuilder.cpp.

What we need/want is to use the simplified text measuring code path in `LayoutTreeBuilder` for simple text, but _still_ use the `ComplexTextController` for measuring. If a TextRun is actually "complex" (if codePath() returns `CodePath::Complex`) we want to use the non-simplified text measuring code path and the `ComplexTextController`.

vitorroriz replied:
Thanks for clarifying it!

nikolaszimmermann commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -458,7 +471,15 @@ inline float FontCascade::widthForTextUsingSimplifiedMeasuring(StringView text,
+#if !ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH)
     return widthForSimpleTextSlow(text, textDirection, cacheEntry);
+#else
+    TextRun run { text, 0, 0, ExpansionBehavior::defaultBehavior(), textDirection, false, false };
```
If we ever see a perf problem with respect to he increased size of the inline function `widthForTextUsingSimplifiedMeasuring` for our ports, we could introduce another `widthForSimpleTextUsingUnifiedComplexCodePath` helper -- for now we deemed this is not necessary.

nikolaszimmermann reviewed (approved):
This looks good to me, r=me.

@vitorroriz since you commented before and had questions/concerns, we'll hold back landing this until you approve it, too. I tried to answer your last question why we're introducing a different `codePathForShaping()` method instead of patching `codePath()` -- please take another look, if you're ok with the approach, or see a better way to achieve what we need :-)

vitorroriz commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -646,7 +646,7 @@ FontCascade::CodePath FontCascade::codePath(const TextRun& run, std::optional<un
         return s_codePath;
 
-#if !USE(FREETYPE)
+#if !USE(FREETYPE) && !ENABLE(COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH)
```
That's interesting. This allows you to use the simple path for partial runs in places that guard against codePath not for shaping. We should test if something like this is possible in the general case as it could be a potential perf progression.

vitorroriz replied:
cc: @Andrew-Fryer

fujii replied:
As far as I know, the simple text code path doesn't support partial text run.
https://github.com/WebKit/WebKit/blob/eb8b277b0314108e5cb2886e39dfc52d98fc845a/Source/WebCore/platform/graphics/FontCascade.cpp#L1447-L1448

vitorroriz replied:
It doesn't because of shaping. I meant something similar like here, to investigate if we could take the simple path for simple text measuring but not for deciding whether to use complex/width "iterators"

fujii committed:
Use ComplexTextController even for the simple text code path for GTK and WPE ports
https://bugs.webkit.org/show_bug.cgi?id=297843

Reviewed by Nikolas Zimmermann and Vitor Roriz.

<https://commits.webkit.org/200076@main> changed GTK and WPE ports to
always use the complex text code path. This was done by using
`FontCascade::setCodePath(CodePath::Complex)`. However, this approach
involved a unnecessary performance regression.

In most of layout perf test cases, the complex and simple text code
paths consume only a tiny portion of time. In WebCore layout and
rendering code, there are some conditions checking the code path like
`if (fontCascade.codePath(TextRun(content)) == FontCascade::CodePath::Complex)`.
This yields the difference in performance. For example, the following
condition in LineBoxBuilder::collectFallbackFonts is a condition
making the difference.

    if (inlineTextBox.canUseSimplifiedContentMeasuring()) {
        // Simplified text measuring works with primary font only.
        return { };
    }

Thus, we shouldn't use `FontCascade::setCodePath()` to choose the
complex text code path.

To use the simple text code path, we have to implement
Font::applyTransforms() for shaping the text. In bug#297088, I studied
the possibility of implementing Font::applyTransforms with HarfBuzz.
However, HarfBuzz doesn't have an API to implement
Font::applyTransforms efficiently.

Use ComplexTextController even for the simple text code path for GTK
and WPE ports. Added a new macro ENABLE_COMPLEX_TEXT_CONTROLLER_FOR_SIMPLE_CODE_PATH.

* LayoutTests/platform/gtk/fast/text/international/pop-up-button-text-alignment-and-direction-expected.txt:
* Source/WTF/wtf/PlatformEnableGlib.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::drawText const):
(WebCore::FontCascade::drawEmphasisMarks const):
(WebCore::FontCascade::displayListForTextRun const):
(WebCore::FontCascade::widthOfTextRange const):
(WebCore::FontCascade::width const):
(WebCore::FontCascade::adjustSelectionRectForText const):
(WebCore::FontCascade::offsetForPosition const):
(WebCore::FontCascade::codePath const):
(WebCore::FontCascade::widthForCharacterInRun const):
(WebCore::FontCascade::lineSegmentsForIntersectionsWithRect const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::widthForTextUsingSimplifiedMeasuring const):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::platformInitialize):

Canonical link: https://commits.webkit.org/299354@main
